QccPack-0.61-1/src/libQccPackAVQ/utils/man1/avqrate.1
.TH AVQRATE 1 "QCCPACK" ""
.SH NAME
avqrate \- calculates rate for AVQ programs
.SH SYNOPSIS
.B avqrate
.RB "[\|" \-vo "\|]"
.RB "[\|" \-noent "\|]"
.I codebook\_coder
.I channelfile
.I sideinfofile
.SH OPTIONS
.TP
.B \-vo
Output only the value of the rate (terse output).
.TP
.B \-noent
Disable entropy-coding estimate in side information.
.SH DESCRIPTION
.LP
.B avqrate
reads a channel,
.I channelfile
(CHN format), and side information,
.I sideinfofile
(SID format), both produced by one of the following AVQ programs:
.BR gtrencode "(1), " paulencode "(1), or " gyencode (1).
It estimates the bit rate, in terms of number of bits per
original source symbol, achieved by the algorithm.
The calculation of the bit rate varies for each algorithm as described below.
.B avqrate
determines which specific calculation to use based on the
.I program\_name 
field specified in the header of
.IR sideinfofile .
.I codebook\_coder
gives the scalar quantizer used to produce the side information.
.SH THE GTR AND PAUL ALGORITHMS
The bit rate is estimated as:
.LP
.I "    R"
=
.RI ( Nc
+
.IR Ns )/ N / dim
.LP
where
.I N
is the total number of original source vectors,
.I dim
is the vector dimension,
.I Nc
is the number of bits used in the VQ channel, and
.I Ns
is the total number of bits of side information.
.I Nc 
is estimated from the first-order entropy of the VQ channel as
.I Nc
= 
.IR Hc *( N " - " Nr ),
where 
.I Hc
is the first-order entropy of the channel and
.I Nr
is the number of times the codebook was updated during
encoding.
.LP
Unless
.B \-noent
is specified,
.I Ns
is estimated as the first-order entropy of the
update flags used by the algorithm plus the first-order
entropy of the components of the update vectors.
If 
.B \-noent
is specified, entropy estimation of side information
is disabled; that is, each update flag counted as one bit 
and a fixed-length block code is assumed for the 
update-vector components.
.LP
.B avqrate
prints
.IR R , 
.IR Nc / N / dim ,
and
.IR Ns / N / dim
to stdout,
unless option
.B \-vo
is given, in which case only
.I R
is output.
.SH THE GERSHO-YANO ALGORITHM
The bit rate is estimated as
.LP
.I "    R"
=
.RI ( Rc
+
.IR Rs )/ dim
.LP
where 
.I dim
is the vector dimension,
.I Rc
is the rate of the VQ channel, and
.I Rs 
is the rate of side information.
.I Rc
is simply the first-order entropy of the VQ channel.
.I Rs
is calculated as
.I Rs
=  
.I Ns
/
.IR N ,
where
.I Ns
is the number of bits of side information and
.I N
is the total number of original source vectors.
.LP
Unless
.B \-noent
is specified,
.I Ns
is estimated as the first-order entropy of the
update addresses used by the algorithm plus the first-order
entropy of the components of the update vectors.
If 
.B \-noent
is specified, entropy estimation of side information
is disabled; that is, a fixed-length block code is assumed for the 
both the update addresses and the update-vector components.
.LP
.B avqrate
prints
.IR R , 
.IR Rc / dim ,
and
.IR Rs / dim
to stdout,
unless option
.B \-vo
is given, in which case only
.I R
is output.
.SH "SEE ALSO"
.BR gtrencode (1),
.BR gtrdecode (1),
.BR paulencode (1),
.BR pauldecode (1),
.BR gyencode (1),
.BR gydecode (1),
.BR QccPackAVQ (3),
.BR QccPack (3)

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackAVQ/utils/avqrate.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "avqrate.h"

#define USG_STRING "[-vo %:] [-noent %:] %s:codebook_coder %s:channelfile %s:sideinfofile"

#define AVQRATE_PROGRAM_GTR   "gtrencode"
#define AVQRATE_PROGRAM_PAUL  "paulencode"
#define AVQRATE_PROGRAM_GY    "gyencode"

QccChannel Channel;
QccAVQSideInfo SideInfo;

int ValueOnly = 0;
int Verbose;
int NoEntropyCodeSideInfo = 0;


int main(int argc, char *argv[])
{
  QccInit(argc, argv);
  
  QccChannelInitialize(&Channel);
  QccAVQSideInfoInitialize(&SideInfo);
  
  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &ValueOnly,
                         &NoEntropyCodeSideInfo,
                         SideInfo.codebook_coder.filename,
                         Channel.filename,
                         SideInfo.filename))
    QccErrorExit();
  
  if (QccSQScalarQuantizerRead(&(SideInfo.codebook_coder)))
    {
      QccErrorAddMessage("%s: Error calling QccSQScalarQuantizerRead()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccChannelReadWholefile(&Channel))
    {
      QccErrorAddMessage("%s: Error calling QccChannelReadWholefile()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccAVQSideInfoStartRead(&SideInfo))
    {
      QccErrorAddMessage("%s: Error calling QccAVQSideInfoStartRead()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (!strlen(SideInfo.program_name))
    QccStringSprintf(SideInfo.program_name, AVQRATE_PROGRAM_GTR);
  
  Verbose = (ValueOnly) ? 1 : 2;

  if (!strcmp(SideInfo.program_name, AVQRATE_PROGRAM_GTR))
    {
      if (QccAVQgtrCalcRate(&Channel, &SideInfo, stdout, NULL, Verbose,
                            !NoEntropyCodeSideInfo))
        {
          QccErrorAddMessage("%s: Error calling QccAVQgtrCalcRate()",
                             argv[0]);
          QccErrorExit();
        }
    }
  else if (!strcmp(SideInfo.program_name, AVQRATE_PROGRAM_PAUL))
    {
      if (QccAVQPaulCalcRate(&Channel, &SideInfo, stdout, NULL,
                             Verbose, !NoEntropyCodeSideInfo))
        {
          QccErrorAddMessage("%s: Error calling QccAVQPaulCalcRate()",
                             argv[0]);
          QccErrorExit();
        }
    }
  else if (!strcmp(SideInfo.program_name, AVQRATE_PROGRAM_GY))
    {
      if (QccAVQGershoYanoCalcRate(&Channel, &SideInfo, stdout, NULL,
                                   Verbose, !NoEntropyCodeSideInfo))
        {
          QccErrorAddMessage("%s: Error calling QccAVQGershoYanoCalcRate()",
                             argv[0]);
          QccErrorExit();
        }
    }
  else
    {
      QccErrorAddMessage("%s: Unrecognized program %s created sideinfo file %s",
                         argv[0], SideInfo.program_name,
                         SideInfo.filename);
    }

  if (QccAVQSideInfoEndRead(&SideInfo))
    {
      QccErrorAddMessage("%s: Error calling QccAVQSideInfoEndRead()",
                         argv[0]);
      QccErrorExit();
    }
  
  QccChannelFree(&Channel);

  QccExit;
}
